jq255.md

```diff
@@ -0,0 +1,1412 @@
+    `m` random bytes obtained from a cryptographically secure source,
+    with a uniform probability distribution (e.g. `/dev/urandom` or the
+    `getrandom()` system call on many Unix-like operating syustems).
+    This is used only for private key generation in the specification of
+    high-level protocols.
+
+If an addition, subtraction or multiplication uses as operands a field
+element and an integer (usually small), then the integer is first
+implicitly converted to a field element as if by using `GFq()`. In
+practice, the implementation may provide a dedicated
+multiply-by-small-integer function; on usual software platforms, such an
+operation is about as fast as an addition in the field.
+
+The implementation of `sqrt(x)` depends on the used field. For jq255s,
+the field is such that `q = 3 mod 4`, which allows a simple square root
+implementation by computing the square root candidate `z = x^((q+1)/4)`.
+For jq255e, the field is such that `q = 5 mod 8`, which can be handled
+with Atkin's algorithm:
+
+ 1. Compute: `c = (2*x)^((q-5)/8)`
+ 2. Compute: `d = 2*x*c^2`
+ 3. Compute the square root candidate: `z = x*c*(d - 1)`
+
+The obtained candidate MUST be checked (by comparing `z^2` with `x` and finding them equivalent).
+Moreover, if `z` is indeed a square root of `x`, then it MUST be
+normalized to a non-negative value (i.e. if `sgn(z) = 1`, then `z` must
+be replaced with `-z`). The use of this fixed convention for square
+roots is needed for proper interoperability of the map-to-group and
+hash-to-group operations.
+
+`is_square(x)` can be implemented as a simple wrapper around `sqrt(x)`,
+though faster implementations are possible with the [Legendre/Jacobi
+symbol]. A fast, constant-time implementation of that operation can be
+obtained as a straightforward derivative of the [optimized binary
+GCD]; on some small microcontrollers, this `is_square()`
+implementation can be about six times faster than the full square root
+computation.
+
+### Affine Formulas
+
+We recall here the affine formulas applicable to points on double-odd
+Jacobi quartic curves. Implementations are expected to use the extended
+representation described later on in this document; the affine formulas
+are potentially useful if some alternate point representations are
+explored.
+
+For any point `P = (e, u)` on the curve, the point `P+N` (i.e. the other
+representant of the same group element) is `P+N = (-e, -u)`. The
+*opposite* of the point `P` is `-P = (e, -u)`. The neutral element of
+the group is `{ I, N }` (i.e. `I = (1, 0)` and `N = (-1, 0)` are the two
+representants of the group neutral).
+
+Let `P1 = (e1, u1)` and `P2 = (e2, u2)` be two points on the curve. Their
+sum `P3 = P1 + P2 = (e3, u3)` can be computed as follows:
+
+```
+     (1 + b'*u1^2*u2^2)*(e1*e2 + a'*u1*u2) + 2*b'*u1*u2*(u1^2 + u2^2)
+e3 = ----------------------------------------------------------------
+                          (1 - b'*u1^2*u2^2)^2
```

Comment:
I think conventions and notations doesn't specify that `/` means multiplication by the inverse.

Reply:
I have added a sentence to that effect.